File: app/src/testUltimate/java/com/clockmods/widget/WidgetPendingIntentFactoryTest.java
package com.clockmods.widget;
import com.clockmods.widget.model.*;
import com.clockmods.widget.render.*;
import com.clockmods.widget.update.*;
import org.junit.Test;
import static org.junit.Assert.*;
public class WidgetPendingIntentFactoryTest {
@Test public void actionsAndInstancesNeverShareIdentity() {
 java.util.Set<Integer> codes=new java.util.HashSet<>(); java.util.Set<String> uris=new java.util.HashSet<>();
 for(int id=1;id<1000;id++) for(WidgetPendingIntentFactory.Action a:WidgetPendingIntentFactory.Action.values()) {
  assertTrue(codes.add(WidgetPendingIntentFactory.requestCode(id,a))); assertTrue(uris.add(WidgetPendingIntentFactory.identity(id,a)));
 }
}
}

File: app/src/testUltimate/java/com/clockmods/widget/WidgetPendingIntentFactoryTest.kt
package com.clockmods.widget

import com.clockmods.ultimate.ComposeMainActivity
import com.clockmods.widget.render.WidgetPendingIntentFactory
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class WidgetPendingIntentFactoryTest {
    @Test
    fun actionsAndInstancesNeverShareIdentity() {
        val codes = HashSet<Int>()
        val uris = HashSet<String>()
        for (id in 1 until 1000) {
            for (action in WidgetPendingIntentFactory.Action.values()) {
                assertTrue(codes.add(WidgetPendingIntentFactory.requestCode(id, action)))
                assertTrue(uris.add(WidgetPendingIntentFactory.identity(id, action)))
            }
        }
    }

    @Test
    fun clockDestinationsMapToExplicitMainRoutes() {
        assertEquals(
            ComposeMainActivity.DESTINATION_CLOCK,
            WidgetPendingIntentFactory.destination(WidgetPendingIntentFactory.Action.CLOCK),
        )
        assertEquals(
            ComposeMainActivity.DESTINATION_CALENDAR,
            WidgetPendingIntentFactory.destination(WidgetPendingIntentFactory.Action.CALENDAR),
        )
        assertNull(WidgetPendingIntentFactory.destination(WidgetPendingIntentFactory.Action.WEATHER))
        assertNull(WidgetPendingIntentFactory.destination(WidgetPendingIntentFactory.Action.CONFIG))
        assertNull(WidgetPendingIntentFactory.destination(WidgetPendingIntentFactory.Action.REFRESH))
    }

    @Test
    fun launcherAcceptsCurrentExtrasAndLegacyWidgetUriActions() {
        assertEquals(
            ComposeMainActivity.DESTINATION_CALENDAR,
            ComposeMainActivity.resolveDestination(ComposeMainActivity.DESTINATION_CALENDAR, null),
        )
        assertEquals(
            ComposeMainActivity.DESTINATION_CLOCK,
            ComposeMainActivity.resolveDestination(null, "CLOCK"),
        )
        assertEquals(
            ComposeMainActivity.DESTINATION_CALENDAR,
            ComposeMainActivity.resolveDestination(null, "CALENDAR"),
        )
        assertNull(ComposeMainActivity.resolveDestination(null, "CONFIG"))
    }
}
